adapters/base/src/main/java/org/apache/cassandra/sidecar/adapters/base/TokenRangeReplicaProvider.java

```diff
@@ -0,0 +1,249 @@
+                                       pendingRangeMappings,
+                                       hostToDatacenter,
+                                       partitioner,
+                                       keyspace);
+
+        Map<String, String> replicaToStateMap = replicaToStateMap(replicaSet, storage);
+
+        return new TokenRangeReplicasResponse(
+        replicaToStateMap,
+        writeReplicas,
+        mappingsToUnwrappedReplicaSet(rangeToEndpointMappings, hostToDatacenter, partitioner));
+    }
+
+    private Map<String, String> replicaToStateMap(Set<String> replicaSet, StorageJmxOperations storage)
+    {
+        List<String> joiningNodes = storage.getJoiningNodesWithPort();
+        List<String> leavingNodes = storage.getLeavingNodesWithPort();
+        List<String> movingNodes = storage.getMovingNodesWithPort();
+
+        String rawGossipInfo = getRawGossipInfo();
+        GossipInfoResponse gossipInfo = GossipInfoParser.parse(rawGossipInfo);
+
+        StateWithReplacement state = new StateWithReplacement(joiningNodes, leavingNodes, movingNodes, gossipInfo);
+
+        return replicaSet.stream()
+                         .collect(Collectors.toMap(Function.identity(), state::of));
+    }
+
+    private String getRawGossipInfo()
+    {
+        return jmxClient.proxy(ClusterMembershipJmxOperations.class, FAILURE_DETECTOR_OBJ_NAME)
+                        .getAllEndpointStatesWithPort();
+    }
+
+    private List<TokenRangeReplicasResponse.ReplicaInfo>
+    writeReplicasFromPendingRanges(Map<List<String>, List<String>> naturalReplicaMappings,
+                                   Map<List<String>, List<String>> pendingRangeMappings,
+                                   Map<String, String> hostToDatacenter,
+                                   Partitioner partitioner,
+                                   String keyspace)
+    {
+        LOGGER.debug("Pending token ranges for keyspace={}, pendingRangeMappings={}", keyspace, pendingRangeMappings);
+        // Merge natural and pending range replicas to generate candidates for write-replicas
+        List<TokenRangeReplicas> replicas = Stream.concat(
+                                                  naturalReplicaMappings.entrySet().stream(),
+                                                  pendingRangeMappings.entrySet().stream())
+                                                  .map(entry -> TokenRangeReplicas.generateTokenRangeReplicas(
+                                                  new BigInteger(entry.getKey().get(0)),
+                                                  new BigInteger(entry.getKey().get(1)),
+                                                  partitioner,
+                                                  new HashSet<>(entry.getValue())))
+                                                  .flatMap(Collection::stream)
+                                                  .collect(Collectors.toList());
```

Comment:
nit: add some indentation to make it easier to read

```suggestion
        List<TokenRangeReplicas> replicas = Stream.concat(naturalReplicaMappings.entrySet().stream(), 
                                                          pendingRangeMappings.entrySet().stream())
                                                  .map(entry -> TokenRangeReplicas.generateTokenRangeReplicas(
                                                      new BigInteger(entry.getKey().get(0)),
                                                      new BigInteger(entry.getKey().get(1)),
                                                      partitioner,
                                                      new HashSet<>(entry.getValue())))
                                                  .flatMap(Collection::stream)
                                                  .collect(Collectors.toList());
```

Reply:
The formatting was not done yet